api/rosetta/metadata.go

```diff
@@ -0,0 +1,100 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+
+	"github.com/onflow/flow-go-sdk"
+
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// MetadataRequest implements the request schema for /construction/metadata.
+// `Options` object in this request is generated by a call to `/construction/preprocess`,
+// and should be sent unaltered as returned by that endpoint.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#request-3
+type MetadataRequest struct {
+	NetworkID identifier.Network `json:"network_identifier"`
+	Options   object.Options     `json:"options"`
+}
+
+// MetadataResponse implements the response schema for /construction/metadata.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#response-3
+type MetadataResponse struct {
+	Metadata object.Metadata `json:"metadata"`
+}
+
+// Metadata implements the /construction/metadata endpoint of the Rosetta Construction API.
+// Metadata endpoint returns information required for constructing the transaction.
+// For Flow, that information includes the reference block and sequence number. Reference block
+// is the last indexed blcok, and is used to track transaction expiration. Sequence number is
+// the proposer account's public key sequence number. Sequence number is incremented for each
+// transaction and is used to prevent replay attacks.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#constructionmetadata
+func (c *Construction) Metadata(ctx echo.Context) error {
+
+	var req MetadataRequest
+	err := ctx.Bind(&req)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidEncoding(invalidJSON, err))
+	}
+
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockchainEmpty))
+	}
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(networkEmpty))
+	}
+
+	if req.Options.AccountID.Address == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(addressEmpty))
+	}
+	if len(req.Options.AccountID.Address) != hexAddressSize {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(addressLength,
+			withDetail("have_length", len(req.Options.AccountID.Address)),
+			withDetail("want_length", hexAddressSize),
+		))
+	}
+
+	current, _, err := c.retriever.Current()
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, internal(txReferenceRetrieval, err))
+	}
+
+	proposer := flow.HexToAddress(req.Options.AccountID.Address)
+	sequenceNr, err := getAccountSequenceNumber(proposer)
```

Comment:
This should already be a function on the retriever interface.

Reply:
Also, is there any other sequence number? If no, why do we keep saying `AccountSequenceNumber` instead of `SequenceNumber`? In fact, I would shorten it even further by using `Nonce`. That's apparently not one hundred percent correct vocabulary, but it is used throughout blockchain projects for this purpose.

Reply:
To be 100% precise, it's the account *key* sequence number of the transaction proposer. Account can have multiple keys and each key has a separate sequence number.

For the time being, I was working under the assumption that it's the first key of an account that's used.

I was planning on iterating on this once the Access API is merged (which is now) and we have the mechanism of retrieving the account and the (key) sequence number.

Do you still think that `Nonce` is an okay term? Something like `ProposerKeyNonce` would be closer to the truth. Not sure if that distinction is important..

Reply:
You are again trying to put every detail about the value into the variable name. That is not necessary. There is nothing ambiguous about `SequenceNumber` or even `Nonce`. It isn't used anywhere else.